src/storage/statistics/variant_stats.cpp

```diff
@@ -0,0 +1,181 @@
+	BaseStatistics::Construct(stats.child_stats[0], VariantShredding::GetUnshreddedType());
+}
+
+void VariantStats::Construct(BaseStatistics &stats) {
+	stats.child_stats = unsafe_unique_array<BaseStatistics>(new BaseStatistics[2]);
+	GetDataUnsafe(stats).is_shredded = false;
+	CreateUnshreddedStats(stats);
+}
+
+BaseStatistics VariantStats::CreateUnknown(LogicalType type) {
+	BaseStatistics result(std::move(type));
+	result.InitializeUnknown();
+	GetDataUnsafe(result).is_shredded = false;
+	result.child_stats[0].Copy(BaseStatistics::CreateUnknown(VariantShredding::GetUnshreddedType()));
+	return result;
+}
+
+BaseStatistics VariantStats::CreateEmpty(LogicalType type) {
+	BaseStatistics result(std::move(type));
+	result.InitializeEmpty();
+	GetDataUnsafe(result).is_shredded = false;
+	result.child_stats[0].Copy(BaseStatistics::CreateEmpty(VariantShredding::GetUnshreddedType()));
+	return result;
+}
+
+BaseStatistics VariantStats::CreateShredded(const LogicalType &shredded_type) {
+	BaseStatistics result(LogicalType::VARIANT());
+	result.InitializeEmpty();
+	auto &variant_stats = GetDataUnsafe(result);
+	variant_stats.is_shredded = true;
+	result.child_stats[0].Copy(BaseStatistics::CreateEmpty(VariantShredding::GetUnshreddedType()));
+	BaseStatistics::Construct(result.child_stats[1], shredded_type);
+	result.child_stats[1].Copy(BaseStatistics::CreateEmpty(shredded_type));
+	return result;
+}
+
+const BaseStatistics &VariantStats::GetShreddedStats(const BaseStatistics &stats) {
+	AssertVariant(stats);
+	D_ASSERT(GetDataUnsafe(stats).is_shredded == true);
+	return stats.child_stats[1];
+}
+
+BaseStatistics &VariantStats::GetShreddedStats(BaseStatistics &stats) {
+	AssertVariant(stats);
+	D_ASSERT(GetDataUnsafe(stats).is_shredded == true);
+	return stats.child_stats[1];
+}
+
+const BaseStatistics &VariantStats::GetUnshreddedStats(const BaseStatistics &stats) {
+	AssertVariant(stats);
+	return stats.child_stats[0];
+}
+
+BaseStatistics &VariantStats::GetUnshreddedStats(BaseStatistics &stats) {
+	AssertVariant(stats);
+	return stats.child_stats[0];
+}
+
+void VariantStats::SetUnshreddedStats(BaseStatistics &stats, const BaseStatistics &new_stats) {
+	AssertVariant(stats);
+	stats.child_stats[0].Copy(new_stats);
+}
+
+void VariantStats::SetUnshreddedStats(BaseStatistics &stats, unique_ptr<BaseStatistics> new_stats) {
+	AssertVariant(stats);
+	if (!new_stats) {
+		CreateUnshreddedStats(stats);
+	} else {
+		SetUnshreddedStats(stats, *new_stats);
+	}
+}
+
+void VariantStats::SetShreddedStats(BaseStatistics &stats, const BaseStatistics &new_stats) {
+	AssertVariant(stats);
+	auto &data = GetDataUnsafe(stats);
+	if (!data.is_shredded) {
+		BaseStatistics::Construct(stats.child_stats[1], new_stats.GetType());
+		data.is_shredded = true;
+	}
+	stats.child_stats[1].Copy(new_stats);
+}
+
+void VariantStats::Serialize(const BaseStatistics &stats, Serializer &serializer) {
+	auto &unshredded_stats = VariantStats::GetUnshreddedStats(stats);
```

Comment:
This should probably also serialize the shredded stats

Reply:
We should add a test that calls `stats` and verifies the stats are (1) printable without crashing, and (2) correct or something

Reply:
Yep properly serializing the shredded stats opened up a whole new can of worms to work through, including the shredded type diversion we were talking about:
```
(lldb) p stats.child_stats[1].type.ToString()
(std::string) "STRUCT(untyped_value_index UINTEGER, typed_value INTEGER)"
(lldb) p other.child_stats[1].type.ToString()
(std::string) "STRUCT(untyped_value_index UINTEGER, typed_value BOOLEAN)"
```